coderd/metricscache/metricscache.go

```diff
@@ -294,26 +317,80 @@ func (c *Cache) Close() error {
 	return nil
 }
 
-func (c *Cache) DeploymentDAUs() (*codersdk.DeploymentDAUsResponse, bool) {
+func (c *Cache) DeploymentDAUs(offset int) (int, *codersdk.DAUsResponse, bool) {
 	m := c.deploymentDAUResponses.Load()
-	return m, m != nil
+	if m == nil {
+		return 0, nil, false
+	}
+	closestOffset, resp, ok := closest(*m, offset)
+	if !ok {
+		return 0, nil, false
+	}
+	return closestOffset, &resp, ok
 }
 
 // TemplateDAUs returns an empty response if the template doesn't have users
 // or is loading for the first time.
-func (c *Cache) TemplateDAUs(id uuid.UUID) (*codersdk.TemplateDAUsResponse, bool) {
+// The cache will select the closest DAUs response to given timezone offset.
+func (c *Cache) TemplateDAUs(id uuid.UUID, offset int) (int, *codersdk.DAUsResponse, bool) {
 	m := c.templateDAUResponses.Load()
 	if m == nil {
 		// Data loading.
-		return nil, false
+		return 0, nil, false
 	}
 
-	resp, ok := (*m)[id]
+	closestOffset, resp, ok := closest(*m, offset)
 	if !ok {
 		// Probably no data.
-		return nil, false
+		return 0, nil, false
+	}
+
+	tpl, ok := resp[id]
+	if !ok {
+		// Probably no data.
+		return 0, nil, false
+	}
+
+	return closestOffset, &tpl, true
+}
+
+// closest returns the value in the values map that has a key with the value most
+// close to the requested key. This is so if a user requests a timezone offset that
+// we do not have, we return the closest one we do have to the user.
+func closest[V any](values map[int]V, offset int) (int, V, bool) {
```

Comment:
What do you mean by that? There are only 24 timezones, and I just used this instead of input validation because idk if there is some odd edge case of a -13 offset.

As for passing it to postgres. We cache these values and reuse them for all calls. So when the caller requests the data, they pull from a cache, they do not trigger a query. So the data must already be available.

I suppose a better cache would populate things lazily, but that would require a larger refactor.

Reply:
Gotcha, I think I misinterpreted your comment as stating that there are 24 hours timezone in the world. I think an hour error tolerance is immaterial, so am good with the implentation you've described. But, I haven't checked the code.

Reply:
Ah, that makes sense. I forgot that everything is cached in the background rather than on-demand. It's unfortunate that we need `count(templates) * 24` queries for each tick though.

Reply:
It is unfortunate, but it runs once per hour, so I am not too worried.